### ISSUE TITLE
Narrow manuscript search/browse by country, city, repository, collection

### ISSUE THREAD

**Issue by nathangibson** [biblia-arabica/srophe-eXist-app#40]
Eventually we will be building a manuscripts database, but for now we want to be able to effectively search manuscripts by their shelfmark. Right now we separate the parts of an identifier with commas in "MS: " tags in Zotero, but I imagine we will need some primitive TEI files containing identifiers like this one in order to do this properly: https://github.com/biblia-arabica/biblia-arabica-data/blob/dev/data/manuscripts/tei/1.xml.

The idea is that in the advanced search and browse pages, we want to be able to narrow down the selection of manuscripts by country, and then cities within that country, repositories within that city, and collections within that repository. A good example is the Advanced Search on the bottom left of this page: https://www.vhmml.org/readingRoom/. As you'll see in the above XML file, these are all separate fields under /TEI/teiHeader/fileDesc/sourceDesc/msDesc/msIdentifier. 

What do you see as our options for this?

**Comment by nathangibson** [comment]
For now I have used an inelegant hack by tokenizing the values in manuscript tags in the format: "MS: City, Collection, Shelfmark". See https://github.com/biblia-arabica/srophe-eXist-app/commit/d9d6dab1ef47b9965063c2576933a00f015fb272

**Comment by nathangibson** [comment]
Later we can build this off individual ms record files.

**Comment by nathangibson** [comment]
We would still like to include this feature if possible, i.e., a drop-down in the Browse filter area that successively filters by Manuscript Location, Collection, and Shelfmark. See the example from [vHMML](https://www.vhmml.org/readingRoom/) when you click in the search bar.

Using as an example [MS: Oxford, Bodleian Libraries, Hunt. 206](https://biblia-arabica.com/bibl/browse.html?fq=;fq-Manuscript%20Shelfmark:Oxford,%20Bodleian%20Libraries,%20Hunt.%20206), the XPath for this would be
1) User sees drop-down of manuscript locations: `/TEI/text/body/listRelation/relation/desc/msDesc/msIdentifier/settlement`
2) User selects "Oxford" > gets list of collections: `/TEI/text/body/listRelation/relation/desc/msDesc/msIdentifier[settlement="Oxford"]/collection`
3) User selects "Bodleian Libraries" > gets list of shelfmarks: `/TEI/text/body/listRelation/relation/desc/msDesc/msIdentifier[settlement="Oxford" and collection="Bodleian Libraries"]/idno[@type="shelfmark"]`

The same thing would be nice to do for Biblical Parts / Sections / Books, but it would be more complex since we would need to either use a list of which parts and sections contain what, or change the way we do the Zotero tagging. But we can figure that out later.

![virtual_hill_museum___manuscript_library](https://user-images.githubusercontent.com/3083452/48192622-9a428000-e348-11e8-8ba2-0a03f67a9286.png)

**Comment by nathangibson** [comment]
Oops, meant to reopen issue with previous commen.t

**Comment by wsalesky** [comment]
@nathangibson Just to clarify you want this in the browse, not the search? As drop down menus? I assume they would replace the facets that are currently in place for these categories?

**Comment by nathangibson** [comment]
Yes, yes, and yes.

**Comment by wsalesky** [comment]
righty-o

**Comment by wsalesky** [comment]
This has been added. @nathangibson @rvollandt please test the feature and make sure it is performing as expected. We can tweak it if need be. 
Thanks.

**Comment by nathangibson** [comment]
Thanks @wsalesky! This is getting toward what we want, but the higher dropdowns should narrow the selection in the lower ones. At this point, you see all collections and all shelfmarks regardless of which city or collection is selected.
City should narrow collections and shelfmarks
Collection should narrow shelfmarks
See https://github.com/biblia-arabica/srophe-eXist-app/issues/40#issuecomment-436948812 for suggested XPath.

**Comment by wsalesky** [comment]
@nathangibson Sorry about that! I changed the names of the facets in the facet definition file and it broke the filter. Should work now. Note, if you do want to change the names/labels of the facets you also have to change them in the facet.xqm file.

**Comment by nathangibson** [comment]
Better :-). When collection is selected, the shelfmarks show the correct selection. 
Still a couple bugs: 
 1. When only city (not collection) is selected, some shelfmarks are displayed that should not be. E.g., if you select Berlin https://biblia-arabica.com/bibl/browse.html?fq=;fq-City:Berlin you see héb. 292, etc. for shelfmark. It seems to be grabbing *all* shelfmarks of records that cite Berlin shelfmarks rather than only the shelfmarks contained within msIdentifier[settlement="Berlin"].
 2. If you've selected a city and then change it, it throws a cardinality error. E.g.., https://biblia-arabica.com/bibl/browse.html?fq=City:Berlin;fq-Collection:BBAW-GCS;fq-City:London. Is it possible to code it so that changing the city drop-down *replaces* the already existing city filter? It seems to be doing this fine when you switch collection or shelfmark.
 3. If you've selected a city and a collection or shelfmark, and then change the collection or shelfmark, the city field goes back to displaying "all" but still filtering by the original city. But then if you try selecting/changing the city from "all" to somehting else, you get the error mentioned in no. 2 above.

**Comment by wsalesky** [comment]
@nathangibson fixed. Let me know if you run across any other quirks. (Also, what the path to the bug was, i.e. what you clicked on.)

Thanks

**Comment by nathangibson** [comment]
@wsalesky 
The bug path was https://biblia-arabica.com/bibl/browse.html?fq=City:Berlin;fq-Collection:BBAW-GCS;fq-City:London 
I clicked on City: Berlin then Collection: BBAW-GCS then changed City to London. 

One other thing. Ideally, "Collection" and "Shelfmark" would be hidden until you select a "City." Then Shelfmark would remain hidden until you select Collection. It already functions this way (intentionally or not) when you select a city -- the shelfmark is then hidden but the collection shows, e.g., https://biblia-arabica.com/bibl/browse.html?fq=City:Amesterdam. 
How doable is that?

**Comment by wsalesky** [comment]
@nathangibson  Done. Collection and shelfmark are hidden until a city is selected, shelfmark remains hidden until a collection is selected.

**Comment by nathangibson** [comment]
Great! But we cannot close it yet since it's the neverending issue :-)
1. Selecting "All" (for City / Collection / Shelfmark) should clear any previous filters for that particular facet  (City / Collection / Shelfmark respectively). Currently switching from a selection to all keeps the filter and also gives no way to clear it. 
2. This just in from the boss: Above "City" can we provide the heading "Filter by manuscript location" ?
Thanks for your patience!

**Comment by wsalesky** [comment]
@nathangibson This is done. The data seems to be missing again, want me to re-run and load it? Any idea what happened?

**Comment by nathangibson** [comment]
Weird, no idea what happened this time. On the eXist server I again see “Sorry, an error occurred” when I navigate to ba-data/data. Do you think our eXist installation is corrupted?

**Comment by nathangibson** [comment]
FYI I reloaded the same ba-data package back onto the server and it's working again. No idea what changed.

**Comment by nathangibson** [comment]
One more thing: Currently changing the City drop-down clears the Collection and Shelfmark facets, as it should. But changing the Collection drop-down should also clear the Shelfmark facet (it does not). 
You can see this if you start with https://biblia-arabica.com/bibl/browse.html?fq=;fq-City:Berlin;fq-Collection:BBAW-GCS;fq-Shelfmark:Akz.-Nr.%20481/27 and change "Collection" to "Staatsbibliothek". You then get https://biblia-arabica.com/bibl/browse.html?fq=;fq-;fq-City:Berlin;fq-Shelfmark:Akz.-Nr.%20481/27;fq-Collection:Staatsbibliothek but you should get https://biblia-arabica.com/bibl/browse.html?fq=;fq-;fq-City:Berlin;fq-Collection:Staatsbibliothek

**Comment by wsalesky** [comment]
@nathangibson I'm so sorry, I thought I tested in locally and remotely! This should now be working as requested.

**Comment by nathangibson** [comment]
Hmm, still getting the results mentioned above. Perhaps a caching issue on my end or eXist?

**Comment by wsalesky** [comment]
Did you do a hard refresh? Or try another browser? I tested it with the link above and it works as expected.

**Comment by nathangibson** [comment]
It looked like it worked once, but I've tried several hard refreshes and it's still behaving unexpectedly. Tried a different example in case the other one was cached. See what this does for you: 
Start with https://biblia-arabica.com/bibl/browse.html?fq=;fq-City:London;fq-Collection:Bible%20Society;fq-Shelfmark:unspecified%20shelfmark
Then change Collection to British Library. 
I get two results and this URL: https://biblia-arabica.com/bibl/browse.html?fq=;fq-;fq-City:London;fq-Shelfmark:unspecified%20shelfmark;fq-Collection:British%20Library
But it should have 123 results like this https://biblia-arabica.com/bibl/browse.html?fq=;fq-;fq-City:London;fq-Collection:British%20Library
